runtime/src/main/cpp/Math.cpp

```diff
@@ -0,0 +1,345 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <float.h>
+#include <math.h>
+#include <stdlib.h>
+
+#include "Exceptions.h"
+#include "Types.h"
+
+#if (__MINGW32__ || __MINGW64__)
+namespace {
+
+    // MinGW's implmenetation if asinh/acosh function returns NaN for large arguments so we use another implementation.
+    // Both implementations derived from boost special math functions and are also used by Kotlin/JVM.
+    // Copyright Eric Ford & Hubert Holin 2001.
+
+    constexpr KDouble LN2 = 0.69314718055994530942;
+    constexpr KDouble SQRT2 = 1.41421356237309504880;
+
+    KDouble taylor_2_bound = sqrt(DBL_EPSILON);
+    KDouble taylor_n_bound = sqrt(taylor_2_bound);
+    KDouble upper_taylor_2_bound = 1.0 / taylor_2_bound;
+    KDouble upper_taylor_n_bound = 1.0 / taylor_n_bound;
+
+    KDouble mingw_asinh(KDouble x) {
+        if (x >= +taylor_n_bound) {
+            if (x > upper_taylor_n_bound) {
+                if (x > upper_taylor_2_bound) {
+                    // approximation by laurent series in 1/x at 0+ order from -1 to 0
+                    return log(x) + LN2;
+                } else {
+                    // approximation by laurent series in 1/x at 0+ order from -1 to 1
+                    return log(x * 2 + (1.0 / (x * 2)));
+                }
+            } else {
+                return log(x + sqrt(x * x + 1));
+            }
+        } else if (x <= -taylor_n_bound) {
+            return -mingw_asinh(-x);
+        } else {
+            // approximation by taylor series in x at 0 up to order 2
+            KDouble result = x;
+            if (fabs(x) >= taylor_2_bound) {
+                // approximation by taylor series in x at 0 up to order 4
+                result -= (x * x * x) / 6;
+            }
+            return result;
+        }
+    }
+
+    KDouble mingw_acosh(KDouble x) {
+        if (x < 1) {
+            return NAN;
+        } else if (x > upper_taylor_2_bound) {
+            // approximation by laurent series in 1/x at 0+ order from -1 to 0
+            return log(x) + LN2;
+        } else if (x - 1 >= taylor_n_bound) {
+            return log(x + sqrt(x * x - 1));
+        } else {
+            KDouble y = sqrt(x - 1);
+            // approximation by taylor series in y at 0 up to order 2
+            KDouble result = y;
+            if (y >= taylor_2_bound) {
+                // approximation by taylor series in y at 0 up to order 4
+                result -= (y * y * y) / 12;
+            }
+            return SQRT2 * result;
+        }
+    }
+}
+#endif
+
+extern "C" {
+
+#ifndef KONAN_WASM
```

Comment:
Usually better avoid inverted condition. Better use #if FEATURE, not #if PLATFORM.
So maybe `#if !KONAN_NO_MATH`  here, as on most platforms math.h will be there.